code/espurna/sensors/SM300D2Sensor.h

```diff
@@ -0,0 +1,249 @@
+            if (index == 6) return MAGNITUDE_HUMIDITY;
+            return MAGNITUDE_NONE;
+        }
+
+        // Current value for slot # index
+        double value(unsigned char index) {
+            if (index == 0) return _co2;
+            if (index == 1) return _ch2o;
+            if (index == 2) return _tvoc;
+            if (index == 3) return _pm25;
+            if (index == 4) return _pm100;
+            if (index == 5) return _temperature;
+            if (index == 6) return _humidity;
+            return 0;
+        }
+
+        // Process sensor UART
+        void tick() {
+            _read();
+        }
+
+    protected:
+
+        // ---------------------------------------------------------------------
+        // Protected
+        // ---------------------------------------------------------------------
+
+        bool _serial_is_hardware() {
+            return (3 == _pin_rx) || (13 == _pin_rx);
+        }
+
+        bool _serial_available() {
+            if (_serial_is_hardware()) {
+                return Serial.available();
+            } else {
+                return _serial->available();
+            }
+        }
+
+        void _serial_flush() {
+            if (_serial_is_hardware()) {
+                return Serial.flush();
+            } else {
+                return _serial->flush();
+            }
+        }
+
+        uint8_t _serial_read() {
+            if (_serial_is_hardware()) {
+                return Serial.read();
+            } else {
+                return _serial->read();
+            }
+        }
+
+        // ---------------------------------------------------------------------
+
+        void _parse() {
+
+            #if SENSOR_DEBUG
+            char tmp[50];
+            for (unsigned char i=0; i<17; i++) sprintf(&tmp[2*i], "%02X", _buffer[i]);
+            DEBUG_MSG("[SENSOR] SM300D2: %s\n", tmp);
+            #endif
+
+            // check second header byte
+            if (_buffer[1] != 0x02) {
+                #if SENSOR_DEBUG
+                    DEBUG_MSG("[SENSOR] SM300D2: Wrong header\n");
+                #endif
+                return;
+            }
+
+            // check crc
+            uint8_t crc = 0;
+            for (unsigned char i=0; i<16; i++) crc += _buffer[i];
+            if (crc != _buffer[16]) {
+                #if SENSOR_DEBUG
+                    DEBUG_MSG("[SENSOR] SM300D2: Wrong CRC\n");
+                #endif
+                return;
+            }
+
+            // CO2
+            _co2 = 256 * _buffer[2] + _buffer[3];
+
+            // CH2O
+            _ch2o = 256 * _buffer[4] + _buffer[5];
+
+            // TVOC
+            _tvoc = 256 * _buffer[6] + _buffer[7];
+
+            // PM 2.5
+            _pm25 = 256 * _buffer[8] + _buffer[9];
+
+            // PM 10
+            _pm100 = 256 * _buffer[10] + _buffer[11];
+
+            // Temperature
+            _temperature = _buffer[12] + (float) _buffer[13] / 10.0;
```

Comment:
buffer[12] is signed btw

ref. aliexpress picture-doc
> when temperature data bit7 of b13 = 1, it means negative temperature, and when bit7 = 0 of b13, it means positive temperature. For example, when b13 = 9Bh, at this time bit7 = 1, it means negative temperature, and the actual temperature is -27 C; if B13 = 1Bh, then bit7 = 0, it means positive temperature, then the actual temperature is 27 C

Reply:
You are right, will fix it.